pkg/streamingpromql/types/stats.go

```diff
@@ -21,10 +21,84 @@ type QueryStats struct {
 	// For example, if a query is running with a step of 30s with a range vector selector with range 45s,
 	// then samples in the overlapping 15s are counted twice.
 	TotalSamples int64
+
+	TotalSamplesPerStep []int64
+
+	EnablePerStepStats bool
+	startTimestamp     int64
+	interval           int64
+}
+
+type stepStat struct {
+	T int64
+	V int64
 }
 
 const timestampFieldSize = int64(unsafe.Sizeof(int64(0)))
 
+func (qs *QueryStats) InitStepTracking(start, end, interval int64) {
```

Comment:
Might be easier to pass the `QueryTimeRange` here, then you can use the various helper methods it provides for getting the number of steps, converting between timestamps and step indices etc.

Reply:
If there is no concern of keeping implementation simular to prometheus, it sounds like a good idea, thanks

Reply:
Everything else has diverged, so we might as well make our lives as easy as possible 🙂